README.md
git-rebranch
============

The `git-rebranch` script is a git extension script for easy branch rebasing.


Quickstart
----------

  1. Define the branch layout in the `.gitrebranch` file:

        master
            common
                feature_1
                feature_2

  2. Run `git rebranch`


Description
-----------

The `git-rebranch` is a git extension script for easy branch rebasing.

Let's say you have a branch called `master` and you are developing
new features in branches called `feature_1` and `feature_2`. You realize that
you need to make a changes to a file used in both feature branches. How do
you make the changes so they are visible in both feature branches?

Usually you make a branch called `common` which is based on `master`. Put the
common changes into this branch and rebase the branches `feature_1` and
`feature_2` onto `common`, so the branch layout looks like this:

        master
          +- common
              +- feature_1
              +- feature_2


However manually rebasing everything can be cumbersome and error-prone. We can
certainly do better.

The `git-rebranch` is designed to automate the rebasing. Just describe the
layout of your branches in a file called `.gitrebranch` in the git root
directory and run `git rebranch`.


Installation
------------

It is recommended to use a virtualenv python installation.

Install `git-rebranch`:

        $ python setup.py install


Format of `.gitrebranch`
------------------------

The file `.gitrebranch` must be stored in the repository root directory.

Each line starts with zero or more white-space characters (`tabs` or `spaces`),
followed with the name of the branch. The level of indentation is used to
determine the branch childs and parents.

In the following example the branch `common` will be rebased onto
`master`. Then the branches `feature_1` and `feature_2` will be
rebased onto `common`.

        master
            common
                feature_1
                feature_2


`git-rebranch` uses `git-rebase` internally. Make sure that you don't use it to
rebase any branches you already published because that would break other
people's history.


Options
-------

#### `--dry-run`

If you specify the `--dry-run` option `git-rebranch` will not perform any
changes, it will just show what would have been done. You can use it to check
whether the branch layout looks good:

    $ git rebranch --dry-run

#### `--continue`

If a conflict occurs during rebasing, you can resolve the conflict and then use
`git-rebranch --continue` to continue where it left off.

#### `--abort`

If a conflict occurs during rebasing and you don't want to resolve conflict, use
`git rebranch --abort` to abort the process. Everything that `git-rebranch` has
done so far will be reverted to the original state - even successfully rebased
branches will be reset to their original SHA1-s.


Licence
-------

See the file LICENCE.

Author
------

- Dan Keder (dan.keder@gmail.com)
- <http://github.com/dankeder/git-rebranch>


CHANGES.md
0.2.0
  * Check that all the required branches exist before doing anything
  * Add the --continue and --abort parameters that can be used to continue or
    abort a failed rebranch
  * More useful --dry-run mode
  * Code split into modules and reorganized
  * In setup.py, get the current version using git describe

0.1.0
  * Everything


gitrebranch/git.py
# Copyright (C) 2012  Dan Keder <dan.keder@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

import os
from subprocess import Popen, PIPE


class GitError(Exception):
    def __init__(self, msg, stderr=None):
        self.msg = msg
        self.stderr = stderr

    def __str__(self):
        return self.msg


class Git(object):
    """ This class allows running git commands.
    """
    def __init__(self):
        pass

    @property
    def rootdir(self):
        """ Get the root directory of the git repo.

        :return: Root path to the git repository.
        """
        process = self._run('rev-parse', '--show-toplevel')
        if process.returncode != 0:
            raise GitError("Failed getting git root directory",
                process.stderr.read())
        return process.stdout.readline().strip()

    @property
    def current_branch(self):
        """ Get the current branch name. If the repo is in detached state return
        None.

        :return: Current branch name or None.
        """
        process = self._run('branch')
        if process.returncode != 0:
            raise GitError("Failed getting current branch")
        for line in process.stdout.readlines():
            line = line.strip()
            if line.startswith('*'):
                branch = line[2:]
                if branch == '(no branch)':
                    return None
                else:
                    return branch

    @property
    def isdirty(self):
        """ Get the repository status - i.e. check if there any uncommited
        changes.

        :return: True if the repository is dirty, False otherwise.
        """
        process = self._run('status', '--porcelain')
        if process.returncode != 0:
            raise GitError("Git status failed",
                process.stderr.read())
        for line in process.stdout.readlines():
            if not line.startswith('??'):
                return True
        return False

    def checkout(self, branch):
        """ Checkout a branch.

        :param: branch: Branch to checkout
        """
        process = self._run('checkout', branch)
        if process.returncode != 0:
            raise GitError("Checkout of branch {0} failed".format(branch),
                    process.stderr.read())

    def get_sha1(self, rev):
        """ Get the SHA1 hash of the specified revision.

        :param: rev: Branch name
        :return: SHA1 of the specified revision
        """
        process = self._run('rev-parse', '--short', rev)
        if process.returncode != 0:
            raise GitError("Failed getting SHA1 hash of '{0}'".format(rev))
        sha1 = process.stdout.readline().strip()
        return sha1

    def reset_hard(self, rev):
        """ Perform a "git reset --hard rev".

        :param: rev: Target revision
        """
        process = self._run('reset', '--hard', rev)
        if process.returncode != 0:
            raise GitError("Failed resetting HEAD to {0}".format(rev))

    def rebase_abort(self):
        """ Perform a "git rebase --abort".
        """
        process = self._run('rebase', '--abort')
        process.wait()

        return (process.returncode, process.stdout.read(), process.stderr.read())

    def rebase_continue(self):
        """ Perform a "git rebase --continue".
        """
        process = self._run('rebase', '--continue')
        process.wait()

        return (process.returncode, process.stdout.read(), process.stderr.read())

    def rebase_in_progress(self):
        """ Return True if the repository is in the middle of a rebase
        operation.
        """
        return os.path.exists(os.path.join(self.rootdir, '.git', 'rebase-merge'))

    def rebase(self, newbase, startrev, endrev):
        """ Rebase the commits between the startrev and endrev onto newbase.

        :param: newbase: Commits onto this revision
        :param: startrev: Start revision
        :param: endrev: End revision
        :return: Tuple containing the return code, stdout and stderr output of
        the git rebase command
        """
        orig_end_sha1 = self.get_sha1(endrev)
        process = self._run('rebase', '--merge', '--onto', newbase,
                startrev, endrev)
        process.wait()

        return (process.returncode, process.stdout.read(), process.stderr.read())

    def _run(self, *args):
        """ Run the 'git' command with the specified arguments and return the
            process object.
        """
        cmd = ['git'] + list(args)
        process = Popen(cmd, stdout=PIPE, stderr=PIPE)
        process.wait()
        return process


# vim: expandtab sw=4


gitrebranch/state.py
# Copyright (C) 2012  Dan Keder <dan.keder@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

import os
import pickle


class RebranchState(object):
    def __init__(self, gitroot):
        self._statefile = os.path.join(gitroot, '.git', 'REBRANCH_STATE')

    def store(self, curbranch, orig_branches, plan):
        with open(self._statefile, 'w') as fh:
            pickle.dump((curbranch, orig_branches, plan), fh)

    def load(self):
        with open(self._statefile, 'r') as fh:
            return pickle.load(fh)

    def clear(self):
        os.unlink(self._statefile)

    def in_progress(self):
        return os.path.exists(self._statefile)


# vim: expandtab sw=4


gitrebranch/config.py
# Copyright (C) 2012  Dan Keder <dan.keder@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

import re
import os


class RebranchConfig(object):
    class _Tree(object):
        def __init__(self, val):
            self.val = val
            self.subtrees = []

        def to_str(self, level):
            s = ''
            if self.val is not None:
                s = "{0}\n".format(self.val)
            for node in self.subtrees:
                s += (" " * 4 * level) + node.to_str(level+1)
            return s

        def add_subtree(self, tree):
            self.subtrees.append(tree)

    def __init__(self, git):
        configfile = os.path.join(git.rootdir, ".gitrebranch")
        with open(configfile, 'r') as fh:
            self.read(fh)

    def read(self, fileobj):
        self._fileobj = fileobj

        stack = [RebranchConfig._Tree(None)]
        last = stack[0]
        for (lineno, level, branchname) in self._parselines():
            tree = RebranchConfig._Tree(branchname)

            if level == len(stack) - 1:    # on the same level
                stack[-1].add_subtree(tree)
                last = tree

            elif level == len(stack):      # on the +1 level
                last.add_subtree(tree)
                stack.append(last)
                last = tree

            elif level < len(stack) - 1:   # on the -n level
                stack = stack[:level+1]
                stack[-1].add_subtree(tree)
                last = tree

            else:
                raise RebranchConfigError("Bad indentation", lineno)

        self.root = stack[0]

    def _parselines(self):
        indentwidth = None;
        lineno = 0
        for line in self._fileobj.readlines():
            lineno += 1
            line = line.rstrip()
            if line == '': next

            match = re.match("([\t ]*)(.+)", line)
            if match is None:
                raise RebranchConfigError("Malformed config file format",
                        lineno)
            indent = len(match.group(1))
            branchname = match.group(2)

            # We guess the indent width from the first indented line found. The
            # indent width of all the following lines must be multiple of it.
            if indent:
                if indentwidth is None:
                    indentwidth = indent
                if indent % indentwidth != 0:
                    raise RebranchConfigError("Wrong indentation",
                            lineno)
                level = indent / indentwidth
            else:
                level = 0
            yield (lineno, level, branchname)

    def to_str(self):
        return self.root.to_str(0)


    def rebase_plan(self):
        """ Create a rebasing plan.

        :return: List of rebasing instructions
        """
        plan = []
        stack = [self.root]
        while stack:
            parent = stack.pop()
            for tree in parent.subtrees:
                parentbranch = parent.val
                childbranch = tree.val
                plan.append((parentbranch, childbranch))
                stack = [tree] + stack
        return plan

    @property
    def branches(self):
        """ Return list of all branches mentioned in the config file.

        :return: List of branch names
        """
        result = []
        stack = [self.root]
        while stack:
            parent = stack.pop()
            for tree in parent.subtrees:
                result.append(tree.val)
                stack = [tree] + stack
        return result


class RebranchConfigError(Exception):
    def __init__(self, msg, line=None):
        self.msg = msg
        self.line = line


# vim: expandtab sw=4


gitrebranch/__init__.py
# Copyright (C) 2012  Dan Keder <dan.keder@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

import sys
import argparse
from xtermcolor.ColorMap import XTermColorMap

from gitrebranch.git import Git, GitError
from gitrebranch.config import RebranchConfig, RebranchConfigError
from gitrebranch.state import RebranchState


cmap = XTermColorMap()


def info(msg, *args):
    msg = msg.format(*args)
    sys.stdout.write(cmap.colorize("{0}\n".format(msg), 0x00ff00))


def error(msg):
    sys.stderr.write(cmap.colorize("{0}\n".format(msg), 0xff0000))


def parse_args():
    parser = argparse.ArgumentParser("git-rebranch")

    group = parser.add_mutually_exclusive_group()
    group.add_argument('--continue', action='store_true',
            help='continue interrupted rebranch',
            required=False, dest='cont')
    group.add_argument('--abort', action='store_true',
            help='abort interrupted rebranch',
            required=False)

    parser.add_argument('--dry-run', action='store_true',
            help='dry-run operation')
    parser.add_argument('--verbose', '-v', action='store_true',
            help='verbose operation')

    args = parser.parse_args()

    return args


def _rebranch(git, curbranch, orig_branches, plan, dry_run):
    state = RebranchState(git.rootdir)
    while plan:
        (parentbranch, childbranch) = plan[0]
        plan = plan[1:]
        orig_branches[childbranch] = git.get_sha1(childbranch)

        state.store(curbranch, orig_branches, plan)

        if parentbranch is not None:
            info("Rebasing {0} onto {1}", childbranch, parentbranch)

            # Check if we are in the middle of a rebase?
            if git.rebase_in_progress():
                error("Rebase in progress")
                sys.exit(1)

            # Rebase commit range 'orig_branches[parentbranch]..childbranch' onto
            # the (new) parentbranch
            if not dry_run:
                (rc, stdout, stderr) = git.rebase(parentbranch, orig_branches[parentbranch], childbranch)
                sys.stdout.write(stdout)
                sys.stderr.write(stderr)
                if rc != 0:
                    error('Rebranching failed.')
                    error('To continue, resolve conflicts and run "git rebranch --continue"')
                    error('To stop rebasing and return everything as it were, run "git rebranch --abort"')
                    sys.exit(1)

    state.clear()


def do_rebranch(args):
    git = Git()
    state = RebranchState(git.rootdir)

    # Check if we were interrupted
    if state.in_progress():
        error("Rebranch is in progress. Use --continue or --abort")
        sys.exit(1)

    # Check if the repo is clean
    if git.isdirty:
        error("Working copy is not clean, aborting.")
        sys.exit(1)

    # Remember the current branch
    curbranch = git.current_branch

    try:
        # Read .gitrebranch
        config = RebranchConfig(git)

        # Check that all the required branches exist
        for branch in config.branches:
            git.get_sha1(branch)

        # Do rebranch
        _rebranch(git, curbranch, {}, config.rebase_plan(), args.dry_run)
    except (GitError, RebranchConfigError) as e:
        error(e)
        sys.exit(1)

    # Checkout the original branch
    git.checkout(curbranch)


def do_rebranch_continue(args):
    git = Git()
    state = RebranchState(git.rootdir)

    if not state.in_progress():
        error("There is no rebranch in progress")
        sys.exit(1)

    # Continue the interrupted rebase
    if git.rebase_in_progress():
        info("Continue rebasing")
        (rc, stdout, stderr) = git.rebase_continue()
        sys.stdout.write(stdout)
        sys.stderr.write(stderr)
        if rc != 0:
            error('Resolve the conflicts and run "git rebranch --continue"')
            error('To stop, run "git rebranch --abort"')
            sys.exit(1)

    # Continue rebasing
    try:
        (curbranch, orig_branches, plan) = state.load()
        _rebranch(git, curbranch, orig_branches, plan, args.dry_run)
    except GitError as e:
        error(e)
        sys.exit(1)

    # Checkout the original branch
    git.checkout(curbranch)


def do_rebranch_abort(args):
    git = Git()
    state = RebranchState(git.rootdir)

    if not state.in_progress():
        error("There is no rebranch in progress")
        sys.exit(1)

    # Abort an eventual rebase
    if git.rebase_in_progress():
        (rc, stdout, stderr) = git.rebase_abort()
        sys.stdout.write(stdout)
        sys.stderr.write(stderr)
        if rc != 0:
            error("Failed rebase --abort")

    # Just to be sure; check if the repo is clean
    if git.isdirty:
        error("Working copy is not clean")
        sys.exit(1)

    # Reset rebased branches to their original revisions
    (curbranch, orig_branches, _) = state.load()
    for (branch, sha1) in orig_branches.items():
        info("Resetting {0} to {1}", branch, sha1)
        git.checkout(branch)
        git.reset_hard(sha1)

    # Checkout the original branch
    git.checkout(curbranch)

    state.clear()


def main():
    args = parse_args()

    if args.cont:
        # Continue an interrupted rebranch
        do_rebranch_continue(args)
    elif args.abort:
        # Abort an interrupted rebranch
        do_rebranch_abort(args)
    else:
        # Normal operation
        do_rebranch(args)


if __name__ == '__main__':
    main()

# vim: expandtab sw=4
